File: mobile/ios/Backplane/SpaceView.swift
import SwiftUI
import WebKit

// A bot's space: the JSON src/mobile/space.bend emits (MSpace.json). Plain
// data; Bend has already kept what it understands and capped every size.
// A button calls act("space", send); an input calls
// act("space-input", send + "\u{1f}" + text).

struct SpaceModel: Decodable {
    let bot: String?
    let title: String
    let blocks: [SpaceBlock]
}

struct SpaceBlock: Decodable {
    let type: String
    let text, label, value, hint, action, placeholder, tone, pct, send: String?
    let permille: Int?
    let items, head: [String]?
    let rows: [[String]]?
    let blocks: [SpaceBlock]?

    private enum Key: String, CodingKey {
        case type, text, label, value, hint, action, placeholder, tone, pct, send, permille, items, head, rows, blocks
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: Key.self)
        type = try c.decode(String.self, forKey: .type)
        text = try c.decodeIfPresent(String.self, forKey: .text)
        label = try c.decodeIfPresent(String.self, forKey: .label)
        // a stat's value is text; a progress bar's is a number (read permille instead)
        value = try? c.decodeIfPresent(String.self, forKey: .value)
        hint = try c.decodeIfPresent(String.self, forKey: .hint)
        action = try c.decodeIfPresent(String.self, forKey: .action)
        placeholder = try c.decodeIfPresent(String.self, forKey: .placeholder)
        tone = try c.decodeIfPresent(String.self, forKey: .tone)
        pct = try c.decodeIfPresent(String.self, forKey: .pct)
        send = try c.decodeIfPresent(String.self, forKey: .send)
        permille = try c.decodeIfPresent(Int.self, forKey: .permille)
        items = try c.decodeIfPresent([String].self, forKey: .items)
        head = try c.decodeIfPresent([String].self, forKey: .head)
        rows = try c.decodeIfPresent([[String]].self, forKey: .rows)
        blocks = try c.decodeIfPresent([SpaceBlock].self, forKey: .blocks)
    }

    // what a button or input sends: the bot's value, or the bare action
    var key: String { send ?? action ?? "" }
}

struct SpaceView: View {
    let space: SpaceModel
    let act: (String, String) -> Void

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 14) {
                if !space.title.isEmpty {
                    Text(space.title).font(.title3.weight(.semibold))
                }
                ForEach(Array(space.blocks.enumerated()), id: \.offset) { _, b in
                    SpaceBlockView(block: b, act: act)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding()
        }
        .scrollDismissesKeyboard(.interactively)
    }
}

private struct SpaceBlockView: View {
    let block: SpaceBlock
    let act: (String, String) -> Void

    var body: some View {
        switch block.type {
        case "row":
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(alignment: .top, spacing: 10) {
                    ForEach(Array((block.blocks ?? []).enumerated()), id: \.offset) { _, b in
                        SpaceBlockView(block: b, act: act)
                    }
                }
            }
        default:
            SpaceLeaf(block: block, act: act)
        }
    }
}

private struct SpaceLeaf: View {
    let block: SpaceBlock
    let act: (String, String) -> Void

    var body: some View {
        switch block.type {
        case "heading":
            Text(block.text ?? "").font(.headline).padding(.top, 4)
        case "text":
            Text(LocalizedStringKey(block.text ?? "")).fixedSize(horizontal: false, vertical: true)
        case "stat":
            VStack(alignment: .leading, spacing: 2) {
                Text(block.label ?? "").font(.caption).foregroundStyle(.secondary)
                Text(block.value ?? "").font(.title2.weight(.semibold).monospacedDigit())
                if let h = block.hint, !h.isEmpty {
                    Text(h).font(.caption2).foregroundStyle(.secondary)
                }
            }
            .padding(12)
            .frame(minWidth: 110, alignment: .leading)
            .background(Color(.secondarySystemBackground), in: .rect(cornerRadius: 12))
        case "progress":
            VStack(alignment: .leading, spacing: 4) {
                HStack {
                    Text(block.label ?? "")
                    Spacer()
                    Text(block.pct ?? "").monospacedDigit().foregroundStyle(.secondary)
                }
                .font(.subheadline)
                ProgressView(value: Double(block.permille ?? 0), total: 1000)
            }
        case "list":
            VStack(alignment: .leading, spacing: 4) {
                ForEach(Array((block.items ?? []).enumerated()), id: \.offset) { _, s in
                    HStack(alignment: .firstTextBaseline, spacing: 8) {
                        Text("•").foregroundStyle(.secondary)
                        Text(s)
                    }
                }
            }
        case "kv":
            VStack(spacing: 6) {
                ForEach(Array((block.rows ?? []).enumerated()), id: \.offset) { _, r in
                    HStack(alignment: .firstTextBaseline) {
                        Text(r.first ?? "").foregroundStyle(.secondary)
                        Spacer(minLength: 12)
                        Text(r.count > 1 ? r[1] : "").multilineTextAlignment(.trailing)
                    }
                    .font(.subheadline)
                }
            }
        case "table":
            SpaceTable(head: block.head ?? [], rows: block.rows ?? [])
        case "button":
            Button(block.label ?? "") { act("space", block.key) }
                .buttonStyle(.bordered)
        case "input":
            SpaceInput(block: block, act: act)
        case "divider":
            Divider()
        case "badge":
            Text(block.text ?? "")
                .font(.caption.weight(.medium))
                .padding(.horizontal, 8).padding(.vertical, 3)
                .foregroundStyle(tint)
                .background(tint.opacity(0.15), in: .capsule)
        case "code":
            ScrollView(.horizontal) {
                Text(block.text ?? "").font(.caption.monospaced()).textSelection(.enabled).padding(10)
            }
            .background(Color(.secondarySystemBackground), in: .rect(cornerRadius: 8))
        default:
            EmptyView()
        }
    }

    private var tint: Color {
        switch block.tone {
        case "ok": .green
        case "warn": .orange
        case "bad": .red
        default: .secondary
        }
    }
}

private struct SpaceTable: View {
    let head: [String]
    let rows: [[String]]

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            Grid(alignment: .leading, horizontalSpacing: 16, verticalSpacing: 6) {
                if !head.isEmpty {
                    GridRow { ForEach(Array(head.enumerated()), id: \.offset) { _, h in Text(h).foregroundStyle(.secondary) } }
                        .font(.caption.weight(.semibold))
                    Divider()
                }
                ForEach(Array(rows.enumerated()), id: \.offset) { _, r in
                    GridRow { ForEach(Array(r.enumerated()), id: \.offset) { _, c in Text(c) } }
                        .font(.subheadline)
                }
            }
        }
    }
}

private struct SpaceInput: View {
    let block: SpaceBlock
    let act: (String, String) -> Void
    @State private var text = ""

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            if let l = block.label, !l.isEmpty {
                Text(l).font(.caption).foregroundStyle(.secondary)
            }
            HStack(spacing: 8) {
                TextField(block.placeholder ?? "", text: $text)
                    .textFieldStyle(.roundedBorder)
                    .submitLabel(.send)
                    .onSubmit(send)
                Button("Send", action: send)
                    .buttonStyle(.bordered)
                    .disabled(text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
            }
        }
    }

    private func send() {
        let t = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.isEmpty { return }
        act("space-input", block.key + "\u{1f}" + t)
        text = ""
    }
}

// A bot's space as a web page (mobile/bots.bend's "page"): the hub serves
// it at url behind a policy that runs no script and loads nothing from the
// network, and the web view runs no script either. n is new when the page
// is. A tapped link or submitted form goes to Bend as ("space-link",
// bot + "\u{1f}" + url), which decides what a "space:" one sends; web
// links open in the browser.
struct SpacePageModel: Decodable {
    let bot, url, n: String
}

struct SpacePage: View {
    let model: AppModel
    let page: SpacePageModel
    @State private var html: String?

    private var url: URL? {
        model.hubURL(page.url, query: [URLQueryItem(name: "n", value: page.n)])
    }

    var body: some View {
        Group {
            if let html {
                PageWeb(html: html) { u in
                    if u.scheme == "space" {
                        model.act("space-link", page.bot + "\u{1f}" + u.absoluteString)
                    } else if u.scheme == "http" || u.scheme == "https" {
                        UIApplication.shared.open(u)
                    }
                }
                .ignoresSafeArea(edges: .bottom)
            } else {
                ProgressView().frame(maxWidth: .infinity, maxHeight: .infinity)
            }
        }
        .task(id: url) {
            guard let u = url, let (d, r) = try? await URLSession.shared.data(from: u),
                  (r as? HTTPURLResponse)?.statusCode == 200, let t = String(data: d, encoding: .utf8) else { return }
            html = t
        }
    }
}

private struct PageWeb: UIViewRepresentable {
    let html: String
    let tapped: (URL) -> Void

    func makeCoordinator() -> Coordinator { Coordinator() }

    func makeUIView(context: Context) -> WKWebView {
        let c = WKWebViewConfiguration()
        c.defaultWebpagePreferences.allowsContentJavaScript = false
        c.websiteDataStore = .nonPersistent()
        let w = WKWebView(frame: .zero, configuration: c)
        w.navigationDelegate = context.coordinator
        w.isOpaque = false
        w.backgroundColor = .clear
        return w
    }

    func updateUIView(_ w: WKWebView, context: Context) {
        context.coordinator.tapped = tapped
        if context.coordinator.shown != html {
            context.coordinator.shown = html
            w.loadHTMLString(html, baseURL: nil)
        }
    }

    final class Coordinator: NSObject, WKNavigationDelegate {
        var shown: String?
        var tapped: ((URL) -> Void)?

        // the page itself (and its #anchors) loads; everything else is a tap
        func webView(_ w: WKWebView, decidePolicyFor a: WKNavigationAction,
                     decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
            guard let u = a.request.url else { return decisionHandler(.cancel) }
            if u.scheme == "about" { return decisionHandler(.allow) }
            decisionHandler(.cancel)
            // a redirect the page makes on its own is not a tap
            if a.navigationType != .other { tapped?(u) }
        }
    }
}

File: mobile/android/app/src/main/java/dev/backplane/mobile/SpaceView.kt
package dev.backplane.mobile

import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import org.json.JSONArray
import org.json.JSONObject

// A bot's space: the JSON src/mobile/space.bend emits (MSpace.json). Plain
// data; Bend has already kept what it understands and capped every size.
// A button calls act("space", send); an input calls
// act("space-input", send + "\u001f" + text).

data class SpaceBlock(
    val type: String, val text: String, val label: String, val value: String, val hint: String,
    val action: String, val placeholder: String, val tone: String, val pct: String, val send: String,
    val permille: Int, val items: List<String>, val head: List<String>, val rows: List<List<String>>,
    val blocks: List<SpaceBlock>,
) {
    // what a button or input sends: the bot's value, or the bare action
    val key: String get() = send.ifEmpty { action }
}

data class SpaceModel(val bot: String, val title: String, val blocks: List<SpaceBlock>)

private fun spaceStrs(a: JSONArray?): List<String> =
    if (a == null) emptyList() else (0 until a.length()).map { a.optString(it) }

private fun spaceRows(a: JSONArray?): List<List<String>> =
    if (a == null) emptyList() else (0 until a.length()).map { spaceStrs(a.optJSONArray(it)) }

private fun spaceBlocks(a: JSONArray?): List<SpaceBlock> =
    if (a == null) emptyList() else (0 until a.length()).mapNotNull { a.optJSONObject(it)?.let(::spaceBlock) }

private fun spaceBlock(o: JSONObject): SpaceBlock = SpaceBlock(
    o.optString("type"), o.optString("text"), o.optString("label"),
    // a stat's value is text; a progress bar's is a number (read permille instead)
    if (o.opt("value") is String) o.optString("value") else "",
    o.optString("hint"), o.optString("action"), o.optString("placeholder"), o.optString("tone"),
    o.optString("pct"), o.optString("send"), o.optInt("permille"),
    spaceStrs(o.optJSONArray("items")), spaceStrs(o.optJSONArray("head")), spaceRows(o.optJSONArray("rows")),
    spaceBlocks(o.optJSONArray("blocks")),
)

fun spaceModel(o: JSONObject): SpaceModel =
    SpaceModel(o.optString("bot"), o.optString("title"), spaceBlocks(o.optJSONArray("blocks")))

@Composable
fun SpaceView(space: SpaceModel, act: (String, String) -> Unit) {
    LazyColumn(
        Modifier.fillMaxWidth(),
        contentPadding = androidx.compose.foundation.layout.PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        if (space.title.isNotEmpty()) {
            item { Text(space.title, style = MaterialTheme.typography.titleLarge) }
        }
        itemsIndexed(space.blocks) { _, b -> SpaceBlockView(b, act) }
    }
}

@Composable
private fun SpaceBlockView(b: SpaceBlock, act: (String, String) -> Unit) {
    if (b.type == "row") {
        Row(Modifier.horizontalScroll(rememberScrollState()), horizontalArrangement = Arrangement.spacedBy(10.dp)) {
            b.blocks.forEach { SpaceLeaf(it, act) }
        }
    } else {
        SpaceLeaf(b, act)
    }
}

@Composable
private fun SpaceLeaf(b: SpaceBlock, act: (String, String) -> Unit) {
    val dim = MaterialTheme.colorScheme.onSurfaceVariant
    when (b.type) {
        "heading" -> Text(b.text, style = MaterialTheme.typography.titleMedium, modifier = Modifier.padding(top = 4.dp))
        "text" -> Text(b.text, style = MaterialTheme.typography.bodyMedium)
        "stat" -> Surface(color = MaterialTheme.colorScheme.surfaceVariant, shape = RoundedCornerShape(12.dp)) {
            Column(Modifier.widthIn(min = 110.dp).padding(12.dp), verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Text(b.label, style = MaterialTheme.typography.labelMedium, color = dim)
                Text(b.value, style = MaterialTheme.typography.headlineSmall, fontWeight = FontWeight.SemiBold)
                if (b.hint.isNotEmpty()) Text(b.hint, style = MaterialTheme.typography.labelSmall, color = dim)
            }
        }
        "progress" -> Column(Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Row(Modifier.fillMaxWidth()) {
                Text(b.label, Modifier.weight(1f), style = MaterialTheme.typography.bodyMedium)
                Text(b.pct, style = MaterialTheme.typography.bodyMedium, color = dim)
            }
            LinearProgressIndicator(progress = { b.permille / 1000f }, modifier = Modifier.fillMaxWidth())
        }
        "list" -> Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
            b.items.forEach { s ->
                Row {
                    Text("•", color = dim)
                    Spacer(Modifier.width(8.dp))
                    Text(s, style = MaterialTheme.typography.bodyMedium)
                }
            }
        }
        "kv" -> Column(Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(6.dp)) {
            b.rows.forEach { r ->
                Row(Modifier.fillMaxWidth()) {
                    Text(r.getOrElse(0) { "" }, style = MaterialTheme.typography.bodyMedium, color = dim)
                    Spacer(Modifier.width(12.dp))
                    Text(r.getOrElse(1) { "" }, Modifier.weight(1f), style = MaterialTheme.typography.bodyMedium, textAlign = TextAlign.End)
                }
            }
        }
        "table" -> SpaceTable(b.head, b.rows)
        "button" -> OutlinedButton(onClick = { act("space", b.key) }) { Text(b.label) }
        "input" -> SpaceInput(b, act)
        "divider" -> HorizontalDivider()
        "badge" -> {
            val tint = when (b.tone) {
                "ok" -> Color(0xFF58C28A)
                "warn" -> Color(0xFFE3B341)
                "bad" -> Color(0xFFF06A6A)
                else -> dim
            }
            Text(
                b.text, color = tint, style = MaterialTheme.typography.labelMedium,
                modifier = Modifier.background(tint.copy(alpha = 0.15f), RoundedCornerShape(50)).padding(horizontal = 8.dp, vertical = 3.dp),
            )
        }
        "code" -> Surface(color = MaterialTheme.colorScheme.surfaceVariant, shape = RoundedCornerShape(8.dp), modifier = Modifier.fillMaxWidth()) {
            Text(
                b.text, fontFamily = FontFamily.Monospace, style = MaterialTheme.typography.bodySmall,
                modifier = Modifier.horizontalScroll(rememberScrollState()).padding(10.dp),
            )
        }
        else -> {}
    }
}

@Composable
private fun SpaceTable(head: List<String>, rows: List<List<String>>) {
    val cols = maxOf(head.size, rows.maxOfOrNull { it.size } ?: 0)
    Column(Modifier.horizontalScroll(rememberScrollState()), verticalArrangement = Arrangement.spacedBy(6.dp)) {
        if (head.isNotEmpty()) {
            Row {
                (0 until cols).forEach { i ->
                    Text(head.getOrElse(i) { "" }, Modifier.width(110.dp), style = MaterialTheme.typography.labelMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant)
                }
            }
            HorizontalDivider(Modifier.width(110.dp * cols))
        }
        rows.forEach { r ->
            Row {
                (0 until cols).forEach { i -> Text(r.getOrElse(i) { "" }, Modifier.width(110.dp), style = MaterialTheme.typography.bodyMedium) }
            }
        }
    }
}

@Composable
private fun SpaceInput(b: SpaceBlock, act: (String, String) -> Unit) {
    var text by rememberSaveable(b.key) { mutableStateOf("") }
    val send = {
        val t = text.trim()
        if (t.isNotEmpty()) {
            act("space-input", b.key + "\u001f" + t)
            text = ""
        }
    }
    Column(Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(4.dp)) {
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            OutlinedTextField(
                text, { text = it }, Modifier.weight(1f), singleLine = true,
                label = if (b.label.isEmpty()) null else ({ Text(b.label) }),
                placeholder = { Text(b.placeholder) },
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Send),
                keyboardActions = KeyboardActions(onSend = { send() }),
            )
            TextButton(onClick = send, enabled = text.isNotBlank()) { Text("Send") }
        }
    }
}
